### <h2>ALY6140_Summer2021_Pragati_Koladiya_Assignment_3</h2>

<p> <b>Week: </b>3</p>
<p> <b>Tittle:</b> Assignment 3 </p>

Copyright (c) 2021
<p> <b>Written By:</b> Pragati Koladiya </p> 

## Import Libraries

In [1]:
#Import necessary libraries
import pandas as pd
import numpy as np

In [2]:
arr = np.arange(24).reshape(4, 6)

## (a) Determine the shape, number of dimensions and type of elements in the numpy array arr

In [3]:
### GRADED

print("\nView Array data : ",arr)
print("\nDetermine the shape : ",arr.shape)
print("\nNumber of dimensions : ",arr.ndim)
print("\nType of elements : ",type(arr.size))


View Array data :  [[ 0  1  2  3  4  5]
 [ 6  7  8  9 10 11]
 [12 13 14 15 16 17]
 [18 19 20 21 22 23]]

Determine the shape :  (4, 6)

Number of dimensions :  2

Type of elements :  <class 'int'>


- The array shape is 4 x 6 determined by using shape 
- Dimention is 2D determined by using function ndim
- Type is integer determined by type() function

## (b) Determine the sum of elements in arr

In [4]:
### GRADED

def arr_sum(arr): 

    sum=0
    for i in arr:
        sum = sum + i
          
    return(sum) 


ans = arr_sum(arr)

print("\nThe Array Value")
print(arr) 
print('=====================') 
print("Sum",ans)


The Array Value
[[ 0  1  2  3  4  5]
 [ 6  7  8  9 10 11]
 [12 13 14 15 16 17]
 [18 19 20 21 22 23]]
Sum [36 40 44 48 52 56]


- Created function to calculate the sum of each element using for loop.
- The sum is stored in the ans variable. 

## Exploring Try Catch Block

In [33]:
def division(a, b):
  #the try statement
  try:
      print(a/b)
  #except block to handel if the error occurs,by throwing zero division exception
  except ZeroDivisionError:
      print("Cannot divide by Zero")
  else:
      print("No Error occured")
  finally:
      print('Value of a', a, 'and b', b)

In [34]:
division(10,0)

Cannot divide by Zero
Value of a 10 and b 0


In [35]:
division(0,0)

Cannot divide by Zero
Value of a 0 and b 0


In [36]:
division(0,5)

0.0
No Error occured
Value of a 0 and b 5


In [37]:
division(10,2)

5.0
No Error occured
Value of a 10 and b 2


In [38]:
a = np.array([-1, 0, 1, 2, 3], dtype=float)
b = np.array([ 0, 0, 0, 2, 2], dtype=float)

-> Trying to divide each array value1 with each associated array value2.

In [39]:
c = np.divide(a, b, out=np.zeros_like(a), where=b!=0)
print(c)

[0.  0.  0.  1.  1.5]


- it does the divide calculation anywhere 'where' b does not equal zero. When b does equal zero, then it remains unchanged from whatever value you originally gave it in the 'out' argument.

## (c) Determine the mean of elements in arr

In [5]:
### GRADED

print("\nThe Array Value")
print(arr) 

# mean of the flattened array 
print("\nMean of Array, axis = None : ", np.mean(arr)) 
    
# mean along the axis = 0 
print("\nMean of Array, axis = 0 : ", np.mean(arr, axis = 0)) 
   
# mean along the axis = 1 
print("\nMean of Array, axis = 1 : ", np.mean(arr, axis = 1))



The Array Value
[[ 0  1  2  3  4  5]
 [ 6  7  8  9 10 11]
 [12 13 14 15 16 17]
 [18 19 20 21 22 23]]

Mean of Array, axis = None :  11.5

Mean of Array, axis = 0 :  [ 9. 10. 11. 12. 13. 14.]

Mean of Array, axis = 1 :  [ 2.5  8.5 14.5 20.5]


- Above print staement shows the mean() along with the different axis values.

## (d) Create an array of the same shape as arr but filled with zeros


In [6]:
### GRADED

arr_zero = np.zeros( (4 ,6) , dtype=np.int64)
print(arr_zero)

[[0 0 0 0 0 0]
 [0 0 0 0 0 0]
 [0 0 0 0 0 0]
 [0 0 0 0 0 0]]


- To create empty or zero array, np.zeros() function is being used.

## (e) Create an array of the same shape as arr but filled with ones

In [7]:
### GRADED

arr_one = np.ones( (4 ,6) , dtype=np.int64)
print(arr_one)

[[1 1 1 1 1 1]
 [1 1 1 1 1 1]
 [1 1 1 1 1 1]
 [1 1 1 1 1 1]]


- Created sample array filled with one by using np.ones().

## (f) Create an array of the same shape as arr but where all elements are squared values

In [8]:
### GRADED

arr_one *= 2
print ("Doubled each element of original array:", arr_one)

Doubled each element of original array: [[2 2 2 2 2 2]
 [2 2 2 2 2 2]
 [2 2 2 2 2 2]
 [2 2 2 2 2 2]]


- To double the array whole array needs to be raised by 2 or powered 2. As we can see 1 is being replced with 2. 

In [9]:
### GRADED

arr *= 2
print ("Doubled each element of original array:", arr)

Doubled each element of original array: [[ 0  2  4  6  8 10]
 [12 14 16 18 20 22]
 [24 26 28 30 32 34]
 [36 38 40 42 44 46]]


- Here whole original array is being doubled.

## (g) Create an array result of shape 4 by 4 resulting from multiplication of arr with transpose(arr). Hint: use np.dot


In [10]:
### GRADED

print ("\nOriginal array:\n", arr)
arr_trans = arr.T
print ("\nTranspose of array:\n", arr_trans)


Original array:
 [[ 0  2  4  6  8 10]
 [12 14 16 18 20 22]
 [24 26 28 30 32 34]
 [36 38 40 42 44 46]]

Transpose of array:
 [[ 0 12 24 36]
 [ 2 14 26 38]
 [ 4 16 28 40]
 [ 6 18 30 42]
 [ 8 20 32 44]
 [10 22 34 46]]


- To transpose the array arrayname.T is being used.

In [11]:
### GRADED

# matrix multiplication
print ("Matrix Multiplication \n {Original Array(arr) X Transpose Array(arr_trans)}:\n\n", arr.dot(arr_trans))

Matrix Multiplication 
 {Original Array(arr) X Transpose Array(arr_trans)}:

 [[  220   580   940  1300]
 [  580  1804  3028  4252]
 [  940  3028  5116  7204]
 [ 1300  4252  7204 10156]]


- Once the array is transposed we need to use the function .dot() to do array matrix multiplication. 

## Working with Pandas Dataframe.

In [17]:
### GRADED
# The following piece of code creates a DataFrame 
import pandas as pd
df = pd.read_csv('https://raw.githubusercontent.com/mwaskom/seaborn-data/master/iris.csv')
df.head(10) #prints the first 10 rows

sepal_length  sepal_width  petal_length  petal_width species
0           5.1          3.5           1.4          0.2  setosa
1           4.9          3.0           1.4          0.2  setosa
2           4.7          3.2           1.3          0.2  setosa
3           4.6          3.1           1.5          0.2  setosa
4           5.0          3.6           1.4          0.2  setosa
5           5.4          3.9           1.7          0.4  setosa
6           4.6          3.4           1.4          0.3  setosa
7           5.0          3.4           1.5          0.2  setosa
8           4.4          2.9           1.4          0.2  setosa
9           4.9          3.1           1.5          0.1  setosa

## (a) Determine the shape of the dataframe

In [3]:
### GRADED

df.shape

(150, 5)

# (b) Print out all the columns of the dataframe

In [14]:
### GRADED

df.columns

Index(['sepal_length', 'sepal_width', 'petal_length', 'petal_width',
       'species'],
      dtype='object')

# (c) Print the 3rd element of the dataframe

In [15]:
 ### GRADED
 
df.iloc[[3]]

sepal_length  sepal_width  petal_length  petal_width species
3           4.6          3.1           1.5          0.2  setosa

# (d) Find the average of 'petal_width' where species is 'virginica'

In [18]:
### GRADED

df_species = pd.DataFrame(df.groupby(df['species'] == 'virginica').mean())

In [19]:
df_species

sepal_length  sepal_width  petal_length  petal_width
species                                                      
False           5.471        3.099         2.861        0.786
True            6.588        2.974         5.552        2.026

In [20]:
df_species['petal_width'].loc[True]

2.026

- First, we have filtered the data frame using group by and using loc[] we have fetched only the mean value of virginica species’ petal width average as 2.026

# (e) Find the maximum of 'sepal_width' where species is 'setosa'

In [21]:
### GRADED

df_species_setosa = df.groupby(df['species'] == 'setosa').max()

In [23]:
df_species_setosa

sepal_length  sepal_width  petal_length  petal_width    species
species                                                                 
False             7.9          3.8           6.9          2.5  virginica
True              5.8          4.4           1.9          0.6     setosa

In [22]:
df_species_setosa['sepal_width'].loc[True]

4.4

- First, we have filtered the data frame using group by and using loc[] we have fetched only the max value of setosa species’ sepal width maximum value as 4.4

# (f) What is the average value of sepal_length



In [ ]:
### GRADED

df['sepal_length'].mean()

5.843333333333335

# (g) What is the maximum value of sepal_width

In [ ]:
### GRADED

df['sepal_width'].max()

4.4

# (h) What is the minimum value of petal_width

In [ ]:
### GRADED

df['petal_width'].min()

0.1

## Reference:

https://note.nkmk.me/en/python-numpy-ndarray-ndim-shape-size/

https://numpy.org/doc/stable/reference/generated/numpy.mean.html

https://stackoverflow.com/questions/54293368/python-pandas-access-to-true-and-false-values-after-groupby-and-size-are-applied

### GRADED: 
Getting started on the capstone project
You should start working on your Capstone Project this week. 
For this week, the task consists of finalizing your selection of a dataset of interest.
 - You can use the links in the supplementary materials to help you find the dataset
 - This dataset should be publicly accessible and available for download
 - Should be of interest to the class
 - The goal is to be asking and answering important questions about the data

#### Open datasets
          
* Resources for open data
  * https://data.boston.gov/
  * https://www.data.gov/
  * https://www.bls.gov/data/
  * http://www.fao.org/statistics/databases/en/
  * http://ec.europa.eu/eurostat/web/main
  * http://census.gov
    


# Questions (GRADED)

# **Question 1 : Shortlist 3 (three) datasets of interest**

**A:** Shortlisted three datasets are as follows,

1) UCI Machine Learning Repository

https://archive.ics.uci.edu/ml/datasets/census+income

2) Heart Disease Data Set

https://archive.ics.uci.edu/ml/datasets/Heart+Disease

3) Student Performance Data Set

https://archive.ics.uci.edu/ml/datasets/Student+Performance

# **Question 2 : For each dataset suggest describe the data (a few sentences) and questions that can be answered using that data (a few sentences)**

A: Dataset Exploration - 

## **1) Cencus Income data set**

https://archive.ics.uci.edu/ml/datasets/census+income

**Dataset Info:** Dataset was collected by Barry Becker from the 1994 Census database. 

The data contains 32k observations and 13 attributes or variables related to the changed census.

***Question:*** Prediction task is to determine whether a person makes over 50K a year.

**Note:** I am intrested in this dataset as the new Covid episode has seen a tremendous measure of individuals who who signed up for the stimulus checks of 1200 USD in America in the wake of losing their jobs. One of the necessities for the beneficiaries is to have their yearly earnings under $75,000, which is all available through their yearly assessment report.

Understanding the likely yearly pay of unfiled taxes people can assist to help the government with making key strides in dealing with them and guaranteeing the size of their bank hold. Consequently, the government would profit immensely from an Machine Learning model that can assist with anticipating a person's pay base on their demographic features.

According to the Inflationcalcualtor.com, 50,000USD in 1994 equals to $87,000 in 2019. So by using the UCI data I would like to built the modle and provide the best perorming model. In future, this model can be used to apply for the current year’s dataset.


## **2) Heart Disease Data Set** 

https://archive.ics.uci.edu/ml/datasets/Heart+Disease

**Dataset Info:** This database contains 76 attributes, but all published experiments refer to using a subset of 14 of them. In particular, the Cleveland database is the only one that has been used by ML researchers to this date. 

***Question:*** Predict the diagnosis of heart disease (angiographic disease status) based on the provided features.

-- Value 0: < 50% diameter narrowing

-- Value 1: > 50% diameter narrowing

## **3) Student Performance Data Set**

https://archive.ics.uci.edu/ml/datasets/Student+Performance

**Dataset Info:** This data approach student achievement in secondary education of two Portuguese schools. The data attributes include student grades, demographic, social and school related features) and it was collected by using school reports and questionnaires. Two datasets are provided regarding the performance in two distinct subjects: Mathematics (mat) and Portuguese language (por).

***Question:*** Predict the students final gread G3 range between 0 to 20. 

-- Value 0: fail 

-- Value 1 to 20: Anything other then 0 is pass